README.rst
PyDiscover
==========

*PyDiscover: Simple Secure and Lightweight Python Service Discovery*

:Codename: ZaRZaner0
:Version: 1.0
:Code: https://github.com/cr0hn/pydiscover
:Issues: https://github.com/cr0hn/pydiscover/issues/
:Python version: Python 3.4 and above
:Author: Daniel Garcia (cr0hn) - @ggdaniel

What's PyDiscover
-----------------

PyDiscover is a simple service discovery client and server, designed with simplicity, performance and security in mind. Instead of implement SSDP protocol or any else, use a very simple mechanism: send to the clients information as a JSON format.

PyDiscover is very flexible and lightweight, and incorporates a cypher mechanism to secure (password based) the communication between server and clients.

Features
--------

- Simple usage.
- Configurable multicast service discovery.
- Password protected access to server info (optional).
- AES encryption (if you defines a password).
- Custom channel definition.
- High server performance, based in the new Python asyncio module.
- Server can spread any information to clients. This information are sent/received as a JSON format.
- Simple configuration file

Install
-------

Install is so easy:

.. code-block:: bash

    # python3.4 -m pip install pydiscover

How it works?
-------------

**Architecture**:

PyDiscover is composed by client and server:

- Server listen for multicast clients request in the port 50000 (by default).
- Clients send requests using a multicast address to the port 50000.

**Virtual Channels (or magic)**:

Client and server must transmit information in the same *virtual channel (or magic)*. The magic is a pre-shared word that server/client known. Only messages with this word will be attended, performing the "virtual channel".

**Hidden mode**:

By default (per security reasons) server runs as **hidden mode**. This is: if server receives a messages without the correct magic or with wrong password, doesn't answer nothing to the client request. If we want that server answer with error message, we'll activate explicitly.

**Securing communication**

We can set a password for the server. When it's set, the information will be sent cyphered using AES to the clients. Only in the clients known the password could be understand the messages.

**Sent/received information**

Server must be started with *-d* param. This param referees to a *.cfg* file. This file must have the format:

.. code-block:: ini

    [DEFAULT]
    services = 10.0.0.1
    net_password = asfi0j9ask123

- A *[DEFAULT]* section.
- Any information as: *key = value*.

The DEFAULT section content will be sent as a JSON format to the clients.

Usage
-----

**Server**

Starting server in port 40000, with a password an the virtual channel is build by word: "askskAls828":

.. code-block:: bash

    # pydiscover-server -p 40000 --password 1238d8KKls_jj -m askskAls828 -d example.cfg

Disablind hidden mode:

.. code-block:: bash

    # pydiscover-server -p 40000 --password 1238d8KKls_jj -m askskAls828 -d example.cfg --disable-hidden

You can see more examples typing:

.. code-block:: bash

    # pydiscover-server -h

**Client**

Connecting to the server with the above configuration:

.. code-block:: bash

    # pydiscover-client -p 40000 --password 1238d8KKls_jj -m askskAls828 -v

Real example
------------

.. image:: https://raw.githubusercontent.com/cr0hn/pydiscover/master/pydiscover/example.jpg

What's new?
-----------

Version 1.0.0
+++++++++++++

- First version released

License
-------

PyDiscover is released under BSD licence.


CHANGELOG
Version 1.0.0
=============

Improvements and fixes
----------------------

- First release.

New features
------------

- First release.

__init__.py
# -*- coding: utf-8 -*-
#
# PyDiscover - https://github.com/cr0hn/pydiscover
#
# Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
# following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the
# following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
# following disclaimer in the documentation and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
# products derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
# INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

pydiscover/utils.py
# -*- coding: utf-8 -*-

import pyaes


def prepare_text(password):
	if len(password) < 32:
		return "%s%s" % (password, "".join("0" for _ in range(32 - len(password))))
	else:
		return password[:32]


def _get_crypter(password):
	try:
		dc = pyaes.AESModeOfOperationCTR(password.encode(errors="ignore"))
	except TypeError:
		dc = pyaes.AESModeOfOperationCTR(password)

	return dc


def crypt(text, password):
	if not password:
		return text.encode(errors="ignore")

	try:
		return _get_crypter(password).encrypt(text)
	except (TypeError, AttributeError):
		return _get_crypter(password).encrypt(text.encode(errors="ignore"))


def decrypt(text, password):
	if not password:
		return text.decode(errors="ignore")

	try:
		return _get_crypter(password).decrypt(text).decode(errors="ignore")
	except (TypeError, AttributeError):
		return _get_crypter(password).decrypt(text.decode())



pydiscover/client.py
import json
import socket
import logging
import argparse
import datetime

from pydiscover.utils import prepare_text, crypt, decrypt

logging.basicConfig(level=logging.INFO, format='[%(levelname)s - Client Discovery ] %(asctime)s - %(message)s')
log = logging.getLogger(__name__)


#
# This code is based in: http://stackoverflow.com/a/21090815
#

class TimeStampException(Exception):
	pass


class PasswordMagicException(Exception):
	pass


class TimeOutException(Exception):
	pass


def discover(magic="fna349fn", port=50000, password=None, timeout=5):
	log.info("Looking for a server discovery")

	# Prepare password
	if password:
		password = prepare_text(password)

	# Build message
	msg = "%s%s" % (magic, datetime.datetime.now().timestamp())

	try:
		# Send discover
		s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # create UDP socket
		s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)  # this is a broadcast socket
		s.sendto(crypt(msg, password), ('<broadcast>', port))
		s.settimeout(timeout)

		data, addr = s.recvfrom(1024)  # wait for a packet
	except socket.timeout:
		log.info("No servers found")

		raise TimeOutException("No servers found")

	msg = decrypt(data, password)

	# Get a correlates response
	if msg.startswith(magic):
		msg_details = msg[len(magic):]

		log.debug("Got service announcement from '%s' with response: %s" % ("%s:%s" % addr, msg_details))

		if msg_details.startswith("#ERROR#"):
			error_details = msg_details[len("#ERROR#"):]

			log.debug("Response from server: %s" % error_details)

			if "timestamp" in error_details:
				raise TimeStampException(error_details)
			elif "password" in error_details:
				raise PasswordMagicException(error_details)
		else:
			undecoded_msg = msg_details[len("#OK#"):]

			# Decode the json
			ok_details = json.loads(undecoded_msg)

			return ok_details, "%s:%s" % addr


def main():
	parser = argparse.ArgumentParser(description='PyDiscover Client',
	                                 formatter_class=argparse.RawTextHelpFormatter)

	# Main options
	parser.add_argument('-m', '--magic', dest="MAGIC", help="preamble for streams.", default="fna349fn")
	parser.add_argument('-p', '--port', dest="PORT", type=int, help="listen port. Default 50000", default=50000)
	parser.add_argument("-v", "--verbosity", dest="VERBOSE", action="count", help="verbosity level: -v, -vv, -vvv.",
	                    default=2)

	gr_options = parser.add_argument_group("more options")

	gr_options.add_argument('-t', '--timeout', dest="TIMEOUT", type=int,
	                        help="timeout to wait for a server. Default 5s",
	                        default=5)
	gr_options.add_argument('--password', dest="PASSWORD", help="server access password. Default None", default=None)

	parsed_args = parser.parse_args()

	# Setting
	log.setLevel(abs(50 - (parsed_args.VERBOSE * 10)))

	# Call server
	try:
		response, server = discover(magic=parsed_args.MAGIC,
		                            port=parsed_args.PORT,
		                            password=parsed_args.PASSWORD,
		                            timeout=parsed_args.TIMEOUT)

		log.info("Discovered server: '%s - Response: \"%s\"" % (server, str(response)))
	except Exception as e:
		log.info(e)

if __name__ == '__main__':
	main()


pydiscover/server.py
import os
import json
import asyncio
import logging
import argparse
import datetime
import configparser

from pydiscover.utils import prepare_text, crypt, decrypt
from socket import gethostbyname, gethostname

logging.basicConfig(level=logging.INFO, format='[ %(levelname)-5s - Service Discovery ] %(asctime)s - %(message)s')
log = logging.getLogger(__name__)


class DiscoverServerProtocol:

	magic = None
	my_ip = None
	password = None
	answer = None
	disable_hidden = False

	def connection_made(self, transport):
		self.transport = transport

	def datagram_received(self, data, addr):
		message = decrypt(data, self.password)

		# Get a correlates query
		if message.startswith(self.magic):

			# Check timestamp and if there's more than 20 second between send -> not response
			try:
				timestamp = float(message[len(self.magic):])
			except TypeError:
				log.debug('Received invalid timestamp in package from %s' % ("%s:%s" % addr))
				if self.disable_hidden:
					self.transport.sendto(crypt("%s#ERROR#%s" % (self.magic, "Invalid timestamp"), self.password), addr)
				return

			# Check if packet was generated before 20 seconds
			if datetime.datetime.fromtimestamp(timestamp) < (datetime.datetime.now() - datetime.timedelta(seconds=20)):
				if self.disable_hidden:
					self.transport.sendto(crypt("%s#ERROR#%s" % (self.magic, "Timestamp is too old"), self.password), addr)
				log.debug('Received outdated package from %s' % ("%s:%s" % addr))
				return

			# Timestamp is correct -> continue
			log.debug('Received %r from %s' % (message, "%s:%s" % addr))

			self.transport.sendto(crypt("%s#OK#%s" % (self.magic, self.answer), self.password), addr)
		else:
			if self.disable_hidden:
				self.transport.sendto(("%s#ERROR#%s" % (self.magic, "Invalid MAGIC or Password")).encode(), addr)
			log.debug('Received bad magic or password from %s:%s' % addr)


def server_discover(answer, magic="fna349fn", listen_ip="0.0.0.0", port=50000, password=None, disable_hidden=False):

	my_ip = gethostbyname(gethostname())

	log.info("Starting discover server")

	# Prepare password
	if password:
		password = prepare_text(password)

	# Load the answer as JSON
	config = configparser.ConfigParser()
	config.read_file(open(answer))
	_answer = json.dumps(dict(config["DEFAULT"]))

	# Setup Protocol
	DiscoverServerProtocol.magic = magic
	DiscoverServerProtocol.my_ip = my_ip
	DiscoverServerProtocol.password = password
	DiscoverServerProtocol.disable_hidden = disable_hidden
	DiscoverServerProtocol.answer = _answer

	# Start running
	loop = asyncio.get_event_loop()

	listen = loop.create_datagram_endpoint(DiscoverServerProtocol,
	                                       local_addr=(listen_ip, port),
	                                       allow_broadcast=True)
	transport, protocol = loop.run_until_complete(listen)

	try:
		loop.run_forever()
	except KeyboardInterrupt:
		pass
	finally:
		log.info("Shutdown server")
		transport.close()
		loop.close()


def main():

	example = """
Examples:

	Put server to listen and return the Service Register Server IP:
	%(name)s -d 10.0.0.1

	Put server listen and return various Services Registers Servers:
	%(name)s -d 10.0.0.1,192.168.1.1

	Increase verbosity:
	%(name)s -vvv -d 10.0.0.1

	Securing channel communication setting a password:
	%(name)s -vvv -d 10.0.0.1 --password "lasi)8sn;k18s7hfalsk"

	Changing channel port:
	%(name)s -vvv -d 10.0.0.1 -p 91882 --password "lasi)8sn;k18s7hfalsk"

	Changing the MAGIC:
	%(name)s -vvv -m Iksjj19k2j -d 10.0.0.1 -p 91882 --password "lasi)8sn;k18s7hfalsk"

	Disabling hidden mode:
	%(name)s -vvv --disable-hidden -m Iksjj19k2j -d 10.0.0.1 -p 91882 --password "lasi)8sn;k18s7hfalsk"

* MAGIC:

The MAGIC is a string that defines an internal "channel". Client and server must known the channel. The server only
responds to the clients with known this MAGIC.

* Hidden mode:

By default, the server doesn't response to the clients with ack with and invalid MAGIC, PASSWORD or messages that was
sent more than 20 seconds before the server receive it. If we disable the hidden mode, server will respond this the
appropriate error to the client.
	""" % dict(name="discovery-server")

	parser = argparse.ArgumentParser(description='PyDiscover Server',
	                                 formatter_class=argparse.RawTextHelpFormatter, epilog=example)

	# Main options
	parser.add_argument('-d', '--discover-info', dest="INFO", help="file with info to send to the clients",
	                    required=True)
	parser.add_argument("-v", "--verbosity", dest="VERBOSE", action="count", help="verbosity level: -v, -vv, -vvv.",
	                    default=3)

	gr_options = parser.add_argument_group("more options")
	parser.add_argument('-m', '--magic', dest="MAGIC", help="preamble for streams.", default="fna349fn")
	gr_options.add_argument('--password', dest="PASSWORD", help="server access password. Default None", default=None)
	parser.add_argument('-p', '--port', dest="PORT", type=int, help="listen port. Default 50000", default=50000)
	parser.add_argument('-l', '--listen', dest="IP", help="listen IP. Default 0.0.0.0", default="0.0.0.0")
	parser.add_argument('--disable-hidden', dest="NO_HIDDEN", action="store_true", help="disable hidden mode", default=False)

	parsed_args = parser.parse_args()

	# Setting
	log.setLevel(abs(50 - (parsed_args.VERBOSE * 10)))

	# Call server
	server_discover(answer=parsed_args.INFO,
	                magic=parsed_args.MAGIC,
	                listen_ip=parsed_args.IP,
	                port=parsed_args.PORT,
	                password=parsed_args.PASSWORD,
	                disable_hidden=parsed_args.NO_HIDDEN)

if __name__ == '__main__':
	main()
